### PR TITLE
Findbugs improvements

### PR DESCRIPTION
**Pull request by grzesiek2010**
#### What has been done to verify that this works as intended?
Nothing I just improved our code a bit adding new findbugs' rules.

#### Why is this the best possible solution? Were any other approaches considered?
We had a few rules that could be unblocked so reviewed them and fixed.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
It doesn't require any testing just a review.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java [hunk 1]
```diff
@@ -185,12 +185,6 @@
         RankingWidgetDialog.RankingListener,
         SaveFormIndexTask.SaveFormIndexListener, LocationClient.LocationClientListener, LocationListener {
 
-    // save with every swipe forward or back. Timings indicate this takes .25
-    // seconds.
-    // if it ever becomes an issue, this value can be changed to save every n'th
-    // screen.
-    private static final int SAVEPOINT_INTERVAL = 1;
```
Comment (shobhitagarwal1612) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:192]: This change was introduced in 1778e150d86ebc722ef27d744f38019b6a7a5220 back in 2012

It seems like this was added as a configurable setting to account for any future issues. Are we sure that this is safe to remove now?

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:192]: It was a final filed with value `1` used just for modulo and `x % 1` always returns 0 so it was redundant.

**Review by shobhitagarwal1612 — APPROVED** [review]

**Comment by grzesiek2010** [comment]
Testing this pr would be difficult. If it fixes only findbugs rules and you @shobhitagarwal1612 reviewed the changes I'll merge it and it will be tested on a daily basis with other prs.

**Comment by zestyping** [comment]
I don't think we should have a goal of turning on every FindBugs warning.  Each warning has a cost and a benefit; for some the net effect is positive, for some the net effect is negative.

If the warning catches real defects, it has a benefit.  Every time the warning breaks the build or forces a developer to spend 5 minutes waiting for another fix-build-run cycle, it has a cost.  If the warning encourages us to make the code awkward or less expressive, that's also a cost.

The potential long-term costs are significant, so we should consider each warning carefully before enabling it.

There are three factors in estimating the value of a FindBugs warning:
  - How severe are the bugs that it catches?
  - How reliably does it detect bugs of its type?  (False positives are nothing but cost; false negatives reduce benefit.)
  - How predictable is the detection behaviour? (If we enable a warning to break the build, that implies requiring developers to write code that does not produce that warning.  In order to pass the check, even a developer capable of avoiding all warnings would need to be able to predict where the warnings will occur.)

In general, I believe we should only allow FindBugs to break the build if it detects something that might actually cause incorrect or faulty behaviour (for precision, I use the term "defect" to mean this).  It should not break the build just to force developers to follow conventions or manually perform optimizations; optimization is the compiler's job.

FindBugs also assigns each warning a priority level (also called a confidence level): Low, Medium, or High.  We probably don't need to treat low-priority warnings as build-stoppers.

Here's my take on which ones we should enable and disable ("strong" or "mild" indicates the strength of my preference):

  - `ES_COMPARING_STRINGS_WITH_EQ`: **Strong enable**. This prohibits string comparisons using the `==` operator.  Comparing strings with `== ` is virtually always a defect.  Detection is reliable because the pattern is very simple.

  - `MS_PKGPROTECT`, `MS_FINAL_PKGPROTECT`: **Mild disable**.  This prohibits public mutable static fields.  The downside is that a public constant can never be an array, for example.  The upside is that it defends against the dangerous practice of reaching inside another class and mutating its static data.  I think it would be noticeable to a reviewer if a PR did something this strange, though it might slip through.  This is in the category of warnings that defend against malicious code, i.e. code written by ODK developers to attack us, which is not our situation.

  - `SBSC_USE_STRINGBUFFER_CONCATENATION`: **Strong disable**.  This prohibits using `+` on strings in a loop.  This warning does not detect a defect; it only forces premature optimization and makes our code slightly harder to read.  Benchmarks show that you have to loop a large number of times before the speed difference is significant (e.g. 50,000+ times before you see a 50% speedup).  We rarely use `+` in a loop that runs such a large number of times that we would notice any improvement.

  - `INT_BAD_REM_BY_1`: **Strong disable**.  This prohibits using `%` with a modulus that can be determined to be 1.  This is not a defect.  This warning also prohibits us from ever using `%` with a constant expression that might ever evaluate to 1, which makes using and adjusting constants more costly.  SpotBugs also considers this a low-priority warning.

  - `NS_NON_SHORT_CIRCUIT`: **Strong disable**.  This warning encourages us to use short-circuit logic.  It does not detect a defect.  It is the low-priority version of `NS_DANGEROUS_NON_SHORT_CIRCUIT`; we should enable that one and disable this one.

  - `RI_REDUNDANT_INTERFACES`: **Mild disable**.  This prohibits mentioning two things in the `implements` clause where one inherits from the other.  This is not a defect, though it is a sloppy practice.

  - `DB_DUPLICATE_SWITCH_CLAUSES`: **Strong disable**.  This prohibits a switch statement from having two cases that contain the same body.  This is not a defect.  It is a conceivable state of affairs for code to pass through a state in which two cases contain the same body, as the code evolves.  When a change is needed in the future, the change might not apply to both cases.  Collapsing the cases is often be a tidy thing to do, but being forced to always collapse the cases is a maintenance cost.  SpotBugs also considers this a low-priority warning.

  - `RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE`: **Strong disable**.  This prohibits null checks against anything that the compiler can determine to be always null or always non-null.  This is not a defect; it is just defensive coding.  Further, the detection behaviour depends on optimization logic, which means that predicting whether this warning will occur entails accurately replicating the optimization logic in your head, a potentially large cognitive cost.  SpotBugs also considers this a low-priority warning.

@yanokwa @cooperka Your thoughts?

**Comment by cooperka** [comment]
I strongly agree with all of Ping's evaluations above; I think some of the changes in this PR are dangerous because they remove defensive and/or intentionally explicit code. It hurts readability when humans are expeted to intuit the full logical path that findbugs intuits. I think these rules should be disabled. @zestyping would you like to file an issue?

**Comment by zestyping** [comment]
@cooperka I made an issue just for the "strong" preference ones: #2946.  Happy to adjust the list per further discussion.

**Pull request merged**